File: python/layer.py
import numpy as np

class Layer:
    def __init__(self, size, prev_layer, is_output):
        self.size = size
        self.prev_layer = prev_layer
        self.is_output = is_output
        self.weights, self.biases, self.values, self.activ_values = [], [], [], []
        self.weight_grad, self.bias_grad, self.activ_grad = [], [], []
        if prev_layer != None:
            self.weights = np.divide(np.random.rand(size, prev_layer.size), size)
            self.biases = np.divide(np.random.rand(size), size)

    def __relu(self, x):
        if x < 0:
            return 0
        else:
            return x

    def __relu_deriv(self, x):
        if x <= 0:
            return 0
        else:
            return 1

    def __softmax(self):
        denom = np.sum(np.exp(self.values))
        return [np.divide(np.exp(x), denom) for x in self.values]

    def calculate(self):
        self.values = np.add(np.matmul(self.weights, self.prev_layer.activ_values), self.biases)
        if not self.is_output:
            self.activ_values = [self.__relu(x) for x in self.values]
        else:
            self.activ_values = self.__softmax()

    def calculate_gradient(self, expected=[]):
        if not self.is_output:
            relu_deriv = [self.__relu_deriv(x) for x in self.values]
            self.bias_grad = np.multiply(relu_deriv, self.activ_grad)
        else:
            self.bias_grad = np.subtract(self.activ_values, expected)
        self.weight_grad = np.outer(self.bias_grad, self.prev_layer.activ_values)
        self.prev_layer.activ_grad = np.matmul(np.transpose(self.weights), self.bias_grad)

    def update(self, eta):
        self.weights = np.subtract(self.weights, np.multiply(eta, self.weight_grad))
        self.biases = np.subtract(self.biases, np.multiply(eta, self.bias_grad))

File: image-generator-experiment/src/app/classes/Layer.ts
import {math} from './mathjs';
import {LayerSaveData} from "./LayerSaveData";

export class Layer {
  weights: number[][];
  biases: number[];
  values: number[];
  activValues: number[];
  weightGradient: number[][];
  biasGradient: number[];
  activGradient: number[];
  size: number;
  isOutput: boolean;
  prevLayer: Layer;

  constructor(size: number, prevLayer: Layer, isOutput: boolean) {
    this.size = size;
    this.isOutput = isOutput;
    this.prevLayer = prevLayer;
    this.activValues = math.zeros(size);
    if (prevLayer !== undefined) {
      this.weights = math.divide(math.random([size, prevLayer.size]), size);
      this.biases = math.divide(math.random([size]), size);
    }
  }

  propagate(): void {
    this.values = math.add(math.multiply(this.weights, this.prevLayer.activValues), this.biases);
    if (!this.isOutput) {
      this.activValues = Array.from(this.values, x => this.relu(x));
    } else {
      this.activValues = this.softmax();
    }
  }

  calculateGradient(expectedValues: number[]): void {
    if (!this.isOutput) {
      const reluDeriv = Array.from(this.values, x => this.reluDeriv(x));
      this.biasGradient = math.dotMultiply(reluDeriv, this.activGradient);
    } else {
      this.biasGradient = math.subtract(this.activValues, expectedValues);
    }
    this.prevLayer.activGradient = math.multiply(math.transpose(this.weights), this.biasGradient);
  }

  update(eta: number): void {
    this.weights = math.subtract(this.weights, math.multiply(eta, this.weightGradient));
    this.biases = math.subtract(this.biases, math.multiply(eta, this.biasGradient));
  }

  softmax(): number[] {
    const denom = math.sum(math.exp(this.values));
    return Array.from(this.values, x => math.divide(math.exp(x), denom));
  }

  relu(x: number): number {
    return (x < 0) ? 0 : x;
  }

  reluDeriv(x: number): number {
    return (x <= 0) ? 0 : 1;
  }

  getLayerSaveData(): LayerSaveData {
    return new LayerSaveData(this.size, this.weights, this.biases);
  }
}
